Push images via ZMQ

tfarago opened the pull request:
### Rationale
So far I have come up with this based on the dicussion [here](https://github.com/ufo-kit/libuca/pull/85), summary:
- it merely adds `handle_push_request` to `ucad` and re-uses the rest for control of the cameras (I don't see the necessity for a completely new `ucad-zmq` server)
- it removes the zmq pushing from the camera class itself because it doesn't belong there
- it enables to use it from concert via `camera = Camera('net')` for control
- it enables us to have native `async`-based `camera.grab_push()`, which can be added as a mixin or whatever else extension of the python camera object
- one can specify how many frames will be grabbed & pushed to remove communication overhead

### Use case
![Picture1-1](https://user-images.githubusercontent.com/2648829/158676016-58555b10-2e4d-4066-a030-70994f1dd80d.png)

tfarago committed:
set_property: accept NULL for string

tfarago committed:
set_property: accept NULL for string

tfarago committed:
Add the possibility to push images via ZMQ

tfarago committed:
Add the possibility to push images via ZMQ

tfarago commented on ucad.c:
```diff
@@ -338,6 +429,67 @@ handle_grab_request (GSocketConnection *connection, UcaCamera *camera, gpointer
+    gint num_sent = 0;
+    static gsize size = 0;
+    static gchar *buffer = NULL;
+    static gpointer context, socket;
```
Initialize to NULL

tfarago commented on ucad.c:
```diff
@@ -338,6 +429,67 @@ handle_grab_request (GSocketConnection *connection, UcaCamera *camera, gpointer
+        }
+    }
+    if (buffer == NULL || size != request_size) {
+        buffer = g_realloc (buffer, request_size);
```
Check return value

tfarago commented on ucad.c:
```diff
@@ -338,6 +429,67 @@ handle_grab_request (GSocketConnection *connection, UcaCamera *camera, gpointer
+                         "zmq socket creation failed: %s\n", zmq_strerror (zmq_errno ()));
+            goto send_error_reply;
+        }
+        if (zmq_bind (socket, "tcp://*:5555") != 0) {
```
Make this a parameter like UCA_NET_DEFAULT_PORT

tfarago commented on ucad.c:
```diff
@@ -184,6 +202,79 @@ serialize_param_spec (GParamSpec *pspec, UcaNetMessageProperty *prop)
+    zmq_retval = zmq_send (socket, header, header_size, ZMQ_SNDMORE);
+    g_free (header);
+    if (zmq_retval <= 0) {
+        g_set_error (error, UCA_CAMERA_ERROR, UCA_CAMERA_ERROR_SEND,
```
Correct error code

tfarago commented on ucad.c:
```diff
@@ -184,6 +202,79 @@ serialize_param_spec (GParamSpec *pspec, UcaNetMessageProperty *prop)
+    }
+
+    if (zmq_send (socket, buffer, pixel_size * width * height, 0) <= 0) {
+        g_set_error (error, UCA_CAMERA_ERROR, UCA_CAMERA_ERROR_SEND,
```
Correct error code

tfarago commented on ucad.c:
```diff
@@ -338,6 +429,67 @@ handle_grab_request (GSocketConnection *connection, UcaCamera *camera, gpointer
+    }
+
+    for (guint i = 0; i < request->num_frames; i++) {
+        uca_camera_grab (camera, buffer, &error);
```
Also check the return value and not error if possible

tfarago replied:
And drop the gotos

tfarago commented on uca-net-protocol.h:
```diff
@@ -57,6 +58,14 @@ typedef struct {
+    gsize num_frames;
+    guint width;
+    guint height;
+    gushort pixel_size;
```
Determine these ourselves?!

matze commented:
Cool, much better!

tfarago committed:
Polish the code a bit

tfarago committed:
Polish the code a bit

tfarago committed:
git: ignore all build directories

tfarago committed:
Fix formatting of memory error

tfarago committed:
Fix formatting of memory error

tfarago committed:
Fix some endpoint mistakes

tfarago committed:
Fix some endpoint mistakes

tfarago committed:
Fix documentation of host

there is no "optional" port.

tfarago committed:
Fix documentation of host

there is no "optional" port.

tfarago committed:
Expose ucad's port as read-only

tfarago committed:
Expose ucad's port as read-only

matze commented:
By the way, are you going to do some benchmarking? I mean just pushing via ZMQ does not make packets fly magically faster over the wire.

tfarago commented:
Of course I will. And sure, I don't expect zmq to be much faster. The point is rather that from python I can tell the server to stuff the images down the wire without actually needing to touch them in python. So far, on my laptop I get 4 GB/s with zmq and 5 GB/s with your streams, but I use python to receive the zmq which also slows things down I guess. I will make all kinds of tests later and example experiments in concert.

tfarago committed:
More debug

tfarago committed:
Merge branch 'master' into uca_net_camera_grab_send

tfarago committed:
fix

tfarago committed:
Enable sending poison pill

which the client requests by setting number of images to 0.

tfarago committed:
Enable stopping the stream

tfarago committed:
protocol: enable gint64 num_images specification

tfarago committed:
protocol: add endpoint registration

tfarago committed:
ucad: implement multiple endpoints

tfarago committed:
zmq: enable other than PUSH sockets

tfarago committed:
ucad: clean up multiple endpoints implementation

tfarago commented on ucad.c:
```diff
@@ -686,10 +719,9 @@ handle_push_request (GSocketConnection *connection, UcaCamera *camera, gpointer
-    if (header) {
-        g_free (header);
-    }
+    g_free (payload);
```
Free also payload->buffer!

tfarago committed:
Remove unused definitions

obsolete after letting client setting the endpoints.

tfarago committed:
SQUASH ucad: use GHashTable instead of GTree

tfarago committed:
SQUASH ucad zmq: enable specification of SNDHWM

sarkarchandan committed:
ucad: avoid integer overflow (#22)

* ucad: Add check for potential integer overflow

* chore: ucad: Use correct type specifier

* ucad: Use long type for number of frames

* chore: Use glib types

* ucad: Restore formatting consistency

tfarago committed:
SQUASH ucad: json use h x w instead of w x h

to make the python-unpacked shape immediately usable with numpy.

tfarago committed:
SQUASH: poison pill

tfarago committed:
Enable removing all zmq endpoints

gabs1234 commented:
@tfarago Are there any plans to merge this branch ? Or are there more tests that need to be carried out ?

MarcusZuber committed:
Adds the properties 'mirror' and 'rotate' to all remote cameras

The properties are send with the image metadata, telling the receiver if the frames should be transformed.
This would be a python implementation for the 'standard' transformation:
def _convert_image(image, mirror: bool, rotate: int):
   if mirror:
      image = np.fliplr(image)
   if rotate:
      image = np.rot90(image, k=rotate)
   return image

Adds the properties mirror and rotate to all remote cameras

The properties are send with the image metadata, telling the receiver if the frames should be transformed.
This would be a python implementation for the 'standard' transformation:
def _convert_image(image, mirror: bool, rotate: int):
   if mirror:
      image = np.fliplr(image)
   if rotate:
      image = np.rot90(image, k=rotate)
   return image

MarcusZuber reviewed (approved):
@tfarago I reviewed your changes and I'm fine with it
Could you have a look on mine?

MarcusZuber committed:
document usage of mirror and rotate properties of ucad-camers

MarcusZuber committed:
Use base-UcaCamera properties for mirror and rotate

To be squashed in the commit introducing mirror and rotate

MarcusZuber commented:
@tfarago The `ucad_zmq_create_image_header()` has a ton of parameters. Should I pass the camera instead and extract them within?

tfarago commented:
> @tfarago The `ucad_zmq_create_image_header()` has a ton of parameters. Should I pass the camera instead and extract them within?

No, the create header runs in the loop and we want to ask for the properties only once for performance.
